ships/0026-shipwright-managed-push.md

```diff
@@ -0,0 +1,149 @@
+<!--
+Copyright The Shipwright Contributors
+
+SPDX-License-Identifier: Apache-2.0
+-->
+
+---
+title: shipwright-managed-push
+authors:
+  - @SaschaSchwarze0
+reviewers:
+  - @HeavyWombat
+  - @GabeMontero
+  - @ImJasonH
+approvers:
+  - @adambkaplan
+creation-date: 2022-01-19
+last-updated: 2022-01-19
+status: implementable
+---
+
+# Make the push operation to the container registry a shipwright-managed operation
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [x] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [docs](/docs/)
+
+## Open Questions [optional]
+
+None
+
+## Summary
+
+This ship describes why and how we provide means for build strategies to store the produced container image as a tarball in ephemeral storage instead of doing the image push. The push will then be performed by a shipwright-managed step.
```

Comment:
pros/cons of opening up to storing in persistent storage where non-shipwright managed entities could access?

the build strategies volume ship https://github.com/shipwright-io/community/blob/main/ships/0022-build-strategy-volumes.md is needed

even if we do not want to broaden this SHIP's non goals .... you have some very precise applications there that might leverage a tar ball on a PV, but I think a catch all citing that specific would be a good inclusion here, including the dependency on SHIP 22

Reply:
I need to think about this. So far, all scenarios I have in mind "just" require the image's tarball to be available for the lifetime of the pod, so an emptyDir is enough. Basically allowing the post processing that I am proposing here in a step following the build strategy steps. This applies to the push to the container registry which this ship is specifically about, but also the follow-on scenarios (image scanning, sbom generation).

Which scenarios do you see that would further need it in a persistent volume?

Reply:
> I need to think about this. So far, all scenarios I have in mind "just" require the image's tarball to be available for the lifetime of the pod, so an emptyDir is enough. Basically allowing the post processing that I am proposing here in a step following the build strategy steps. This applies to the push to the container registry which this ship is specifically about, but also the follow-on scenarios (image scanning, sbom generation).
> 
> Which scenarios do you see that would further need it in a persistent volume?

A combination of a couple of scenarios or concepts even, at least how we traditionally consumed them in RH/OpenShift.

Executive summary: features for reducing / minimizing the security risk footprint when building software.

Some details:
- the notion we called out in OpenShift docs as "External Artifacts" .... i.e. https://docs.openshift.com/container-platform/4.9/cicd/builds/creating-build-inputs.html#builds-using-external-artifacts_creating-build-inputs.  Now you'll see in that example the external artifact is obtained via `wget` but in theory it could be retrieved from a shared PV that multiple shipwright builds have access to, which is really just a building block concept for this next bullet ....
- builds in disconnected environments is also a thing with OpenShift customers, and even within Red Hat itself, were they have everything a build could need in a mirrored registry within the network constrained environment, where there is no risk of "contamination" by interacting with the outside world while building you product, or or "external attack" to the env where your software is built .  In fact, the productized form of OpenShift and most if not all RH products are built in a "vacuumed sealed" :-) disconnected network environment, as part of RH's guarantee that no corruptions could be introduced.

Reply:
Okay, still searching for the relation to this PR ... your first scenario is about build inputs. I am very much in line that we in Shipwright must eventually support to get the source through a persistent volume rather than downloading it. This is imo a prereq even for things like a Shipwright Build as step of a Tekton pipeline. This PR is about the build output = the container image. If we think that - in such an embedded step scenario, steps after the Build will want to look at the image and prefer to get it locally rather than need to download it again, then yes, this is a possible future extension.

In the disconnected environment, I guess that the Build would be pushing to the internal registry = define this in the output (or eventually through some Image reference)? For me atm independent of whether the strategy or a Shipwright managed step performs the push operation.

Reply:
For me, the relation to this SHIP/PR boils down to what you've called out in the summary about where the tarball is stored.  You explicitly call out ephemeral storage, and ephemeral storage only.  That is what jumped out at me when I first read it.

If this SHIP took on the added bit that ephemeral storage is one of a choice of storage options for the tarball, it will facilitate future scenarios, like the ones I articulated.  If those did not resonate, maybe another scenario would help.  Consider "chains" of shipwright builds.  Independent of disconnected etc.  i.e BuildStrategy etc. A produces the tarbarll.  BuildStrategy B etc. consume the tarball as an input.  A is essentially a prereq for B.  Both sets builds/buildruns use the same PV.  But we are talking separate Pods in the end for A and B.

But if that scenarios does not resonate with you, or you just want to keep it as simple as it needs to be your the scenario you do cite, I won't resist on deferring, and NOT broadening this SHIP and the accompanying implementation effort to allow for choices in the volume type used for the storing the tarball.  I understand the desire for precise, focused requirements.  It can be taken on with a future SHIP.

But how the SHIP-22 implementation has progressed also gives me some pause here.

Implementation of the Build Strategy Volume stuff has had to wait, where we first had to deprecate our use of Tekton Workspace Emptydir volumes with https://github.com/shipwright-io/build/pull/941 before we could fully start on SHIP-22.

The situations may not be exact here, but there were enough commonalities to give me pause and raise this point.  And I'd like to avoid that if possible.

How about this:
- if you want to add the ability to select the volume type, grown beyond just ephemeral volumes, in this PR, great.
- if not, I would ask for minimally a response here, and ideally some verbiage in the SHIP as well, that expanding beyond just ephemeral storage for storing the tarball is purely an additive exercise, where no deprecation cycles will be needed on what you articulate below in the API changes section.

WDYT?

Reply:
> A produces the tarbarll. BuildStrategy B etc. consume the tarball as an input. A is essentially a prereq for B. Both sets builds/buildruns use the same PV. But we are talking separate Pods in the end for A and B.

Thanks Gabe, that's an interesting scenario I now understand. :-) Though, it has larger implications because then the BuildRun A would likely only produce the tarball and not push an image at all?

So, following suggestion: I think what you propose will be valuable for potential future scenarios. But, I would want to keep the scope of this ship simple for now (I had to slow down myself to not include the image scanning and signing right away ;-) ). Though, I do not think I am putting in things that will be problematic in the future. So, I will mention this potential future scenario and will also mention that there are scenarios where the implicit emptyDir might become a flexible user choice to support those scenarios.

Reply:
> > > A produces the tarbarll. BuildStrategy B etc. consume the tarball as an input. A is essentially a prereq for B. Both sets builds/buildruns use the same PV. But we are talking separate Pods in the end for A and B.
> > 
> > 
> > Thanks Gabe, that's an interesting scenario I now understand. :-) Though, it has larger implications because then the BuildRun A would likely only produce the tarball and not push an image at all?
> 
> Quite possibly yes. Though we have some cases internally where a component produces both an image and an installable binary (where are the bits/files are collected in a tarball), to provide flexibility in how their users install and consume their "component". In fact, the "both" permutation might just be more prevalent.

But I don't have precise data on that ^^ though :-)

There is also desire from image builders wrt depending on other images (and the image layering that can ensue that can impact your image's size) vs. "more efficient" pulling in of dependencies by just copying in a binary (and not other parts of an image potentially), an other permutations around those lines, but we can dive into that with that future SHIP ;-)